servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DelegatingSslConfig.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.netty;
+
+import io.servicetalk.transport.api.SslConfig;
+import io.servicetalk.transport.api.SslProvider;
+
+import java.io.InputStream;
+import java.util.List;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.TrustManagerFactory;
+
+import static java.util.Objects.requireNonNull;
+
+abstract class DelegatingSslConfig<T extends SslConfig> implements SslConfig {
+    private final T delegate;
+
+    DelegatingSslConfig(final T delegate) {
+        this.delegate = requireNonNull(delegate);
+    }
+
+    protected T delegate() {
+        return delegate;
+    }
+
+    @Nullable
+    @Override
+    public TrustManagerFactory trustManagerFactory() {
+        return delegate.trustManagerFactory();
+    }
+
+    @Nullable
+    @Override
+    public Supplier<InputStream> trustCertChainSupplier() {
```

Comment:
Is it expected that these would return a fresh InputStream for each invocation or the same stream? Trying to decide if config is one-shot or reusable and it is unclear.

Reply:
I assume you are asking about the base `SslConfig` interface? Since ownership is transferred (e.g. caller must invoke `close()`) it must be a new instance on each invocation of the `Supplier`. Currently these methods will be invoked at most 1 time per ssl setup. Do you feel the ownership transfer or other impacts should be made more explicit in the javadocs?

Reply:
I would expect SslConfig to be reused for many requests as it seems expensive (reading keys and certs) to construct. The test example shows it being created inline which is dangerous in that we don't provide other examples and someone may copy it and use it in that way in their code where they should be creating one and reuse it many times. (The peril of using tests as sample code).

Otherwise, the other part is whether a "currying" Supplier is permitted, that can return the same stream instance every time or whether a true supplier that returns a "fresh" stream each time is required.  That a Supplier is used rather than providing the stream directly, or something else like contents of the stream might imply that the supplier is invoked more than once. The docs about the stream being closed by the impl are good and helpful but since close() is idempotent and it is not an error to call close many times it doesn't really answer whether the Supplier needs to be reusable.

Reply:
> reused for many requests ... doesn't really answer whether the Supplier needs to be reusable.

The config object will be re-used for each ssl setup/handshake (e.g. connection initialization) and I'll clarify in the comments.

> we don't provide other examples

There are a few examples independent from unit tests which demonstrate `SslConfig` usage (`Http*MutualTLS`, `Http*WithAlpn`), and inline creation is expected to be safe.